src/daemon/auth/common/session.rs

```diff
@@ -0,0 +1,340 @@
+/// expiration, that is handled separately by the AuthProvider.
+pub struct LoginSessionCache {
+    cache: RwLock<HashMap<Token, CachedSession>>,
+    encrypt_fn: fn(&[u8], &[u8], &mut [u8]) -> KrillResult<Vec<u8>>,
+    decrypt_fn: fn(&[u8], &[u8], &[u8]) -> KrillResult<Vec<u8>>,
+    ttl_secs: u64,
+}
+
+impl Default for LoginSessionCache {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl LoginSessionCache {
+    pub fn new() -> Self {
+        LoginSessionCache {
+            cache: RwLock::new(HashMap::new()),
+            encrypt_fn: crypt::encrypt,
+            decrypt_fn: crypt::decrypt,
+            ttl_secs: MAX_CACHE_SECS,
+        }
+    }
+
+    pub fn with_ttl(self, ttl_secs: u64) -> Self {
+        LoginSessionCache {
+            cache: self.cache,
+            encrypt_fn: self.encrypt_fn,
+            decrypt_fn: self.decrypt_fn,
+            ttl_secs,
+        }
+    }
+
+    pub fn with_encrypter(self, encrypt_fn: fn(&[u8], &[u8], &mut [u8]) -> KrillResult<Vec<u8>>) -> Self {
+        LoginSessionCache {
+            cache: self.cache,
+            encrypt_fn: encrypt_fn,
+            decrypt_fn: self.decrypt_fn,
+            ttl_secs: self.ttl_secs,
+        }
+    }
+
+    pub fn with_decrypter(self, decrypt_fn: fn(&[u8], &[u8], &[u8]) -> KrillResult<Vec<u8>>) -> Self {
+        LoginSessionCache {
+            cache: self.cache,
+            encrypt_fn: self.encrypt_fn,
+            decrypt_fn: decrypt_fn,
+            ttl_secs: self.ttl_secs,
+        }
+    }
+
+    fn time_now_secs_since_epoch() -> KrillResult<u64> {
+        Ok(SystemTime::now()
+            .duration_since(UNIX_EPOCH)
+            .map_err(|err| Error::Custom(format!("Unable to determine the current time: {}", err)))?
+            .as_secs())
+    }
+
+    fn lookup_session(&self, token: &Token) -> Option<ClientSession> {
+        match self.cache.read() {
+            Ok(readable_cache) => {
+                if let Some(cache_item) = readable_cache.get(&token) {
+                    return Some(cache_item.session.clone());
+                }
+            }
+            Err(err) => warn!("Unexpected session cache miss: {}", err),
+        }
+
+        None
+    }
+
+    fn cache_session(&self, token: &Token, session: &ClientSession) {
+        match self.cache.write() {
+            Ok(mut writeable_cache) => match Self::time_now_secs_since_epoch() {
+                Ok(now) => {
+                    writeable_cache.insert(
+                        token.clone(),
+                        CachedSession {
+                            evict_after: now + self.ttl_secs,
+                            session: session.clone(),
+                        },
+                    );
+                }
+                Err(err) => warn!("Unable to cache decrypted session token: {}", err),
+            },
+            Err(err) => warn!("Unable to cache decrypted session token: {}", err),
+        }
+    }
+
+    pub fn encode(
```

Comment:
This type, `LoginSessionCache`, is a sort of user session store. That it decodes on the fly and the decoded sessions are short lived are implementation details. I wonder then whether with the right trait and naming we could use that trait as a facade behind which we could later swap the implementation with say an on-disk db or a distributed key-value store etc?